rust/cargo-psibase/src/link.rs

```diff
@@ -535,17 +535,115 @@ pub fn link_module(source : &Module, dest : &mut Module) -> Result<(), anyhow::E
+        let validator = wasmparser::Validator::new_with_features(wasmparser::WasmFeatures{
+            sign_extension: true,
+            bulk_memory: true,
+            simd: true,
+            ..wasmparser::WasmFeatures::default()
+        });
+        return validator;
+    }
+
+    struct ImportFunction {
+        id : FunctionId,
+        name : String,
+        module : String,
+    }
+    impl fmt::Display for ImportFunction {
+        fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+            write!(f, "{}::{}", self.module, self.name)
+        }
+    }
+
+    fn get_import_funcs(module : &Module) -> Vec<ImportFunction> {
+        module.imports.iter().filter_map(|import| {
+            let dest_func = match import.kind {
+                walrus::ImportKind::Function(import_fid) => module.funcs.get(import_fid),
+                _ => return None,
+            };
+            
+            let import = module.imports.get_imported_func(dest_func.id()).unwrap();
+            
+
+            Some(ImportFunction {
+                    id : dest_func.id(),
+                    name: import.name.to_owned(),
+                    module: import.module.to_owned()
+                }
+            )
+        }).collect()
+    }
+
+    #[test]
+    fn imports_removed() -> Result<(), Error> {
+        let (original, polyfill, filled) = setup(SIMPLE_WASM)?;
+
+        let targets_before_fill: Option<Vec<PolyfillTarget>> = get_polyfill_targets(&polyfill, &original)?;
+        let _import_fids : Vec<FunctionId> = if let Some(targets) = targets_before_fill {
+            assert_eq!(targets.len(), 4, "Before fill, polyfill targets should be 4.");
+            targets.into_iter().map(|t| t.dest_fid).collect()
+        } else {
+            vec![]
+        };
+
+        let targets_after_fill: Option<Vec<PolyfillTarget>> = get_polyfill_targets(&polyfill, &filled)?;
+        if let Some(targets) = targets_after_fill {
+            assert_eq!(targets.len(), 0, "After fill, polyfill targets should be 0.");
+        }
+
+        let import_funcs = get_import_funcs(&filled);
+        for f in import_funcs {
+            println!("{}", f);
+        }
+
+        Ok(())
+    }
+
+    #[test]
+    fn valid_wasm_produced_1() -> Result<(), Error> {
+        let (mut original, _, mut filled) = setup(SIMPLE_WASM)?;
+
+        new_validator().validate_all(&original.emit_wasm()).map_err(|e|anyhow!("[Validating original, simple] {}", e))?;
+        new_validator().validate_all(&filled.emit_wasm()).map_err(|e|anyhow!("[Validating filled, simple] {}", e))?;
+        Ok(())
+    }
+
+    #[test]
+    fn valid_wasm_produced_2() -> Result<(), Error> {
+        let (mut original, _, mut filled) = setup(INTERMEDIATE_WASM)?;
+
+        new_validator().validate_all(&original.emit_wasm()).map_err(|e|anyhow!("[Validating original, intermediate] {}", e))?;
+        new_validator().validate_all(&filled.emit_wasm()).map_err(|e|anyhow!("[Validating filled, intermediate] {}", e))?;
```

Comment:
This line fails with this error:
https://hotg-ai.github.io/rune/crate-docs/src/wasmparser/operators_validator.rs.html#323

Reply:
Fixed. Structured control instructions (if, block, etc) were using the wrong type